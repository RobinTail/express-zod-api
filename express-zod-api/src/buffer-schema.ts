import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
