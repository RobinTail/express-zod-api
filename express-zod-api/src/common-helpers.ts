import { Request } from "express";
import * as R from "ramda";
import { z } from "zod";
import { CommonConfig, InputSource, InputSources } from "./config-type";
import { contentTypes } from "./content-type";
import { OutputValidationError } from "./errors";
import { metaSymbol } from "./metadata";
import { AuxMethod, Method } from "./method";

/** @desc this type does not allow props assignment, but it works for reading them when merged with another interface */
export type EmptyObject = Record<string, never>;
export type EmptySchema = z.ZodObject<EmptyObject, "strip">;
export type FlatObject = Record<string, unknown>;

/** @link https://stackoverflow.com/a/65492934 */
type NoNever<T, F> = [T] extends [never] ? F : T;

/**
 * @desc Using module augmentation approach you can specify tags as the keys of this interface
 * @example declare module "express-zod-api" { interface TagOverrides { users: unknown } }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- augmentation
export interface TagOverrides {}
export type Tag = NoNever<keyof TagOverrides, string>;

/** @see https://expressjs.com/en/guide/routing.html */
export const routePathParamsRegex = /:([A-Za-z0-9_]+)/g;
export const getRoutePathParams = (path: string): string[] =>
  path.match(routePathParamsRegex)?.map((param) => param.slice(1)) || [];

const areFilesAvailable = (request: Request): boolean => {
  const contentType = request.header("content-type") || "";
  const isUpload = contentType.toLowerCase().startsWith(contentTypes.upload);
  return "files" in request && isUpload;
};

export const defaultInputSources: InputSources = {
  get: ["query", "params"],
  post: ["body", "params", "files"],
  put: ["body", "params"],
  patch: ["body", "params"],
  delete: ["query", "params"],
};
const fallbackInputSource: InputSource[] = ["body", "query", "params"];

export const getActualMethod = (request: Request) =>
  request.method.toLowerCase() as Method | AuxMethod;

export const getInput = (
  req: Request,
  userDefined: CommonConfig["inputSources"] = {},
): FlatObject => {
  const method = getActualMethod(req);
  if (method === "options") return {};
  return (
    userDefined[method] ||
    defaultInputSources[method] ||
    fallbackInputSource
  )
    .filter((src) => (src === "files" ? areFilesAvailable(req) : true))
    .reduce<FlatObject>((agg, src) => Object.assign(agg, req[src]), {});
};

export const ensureError = (subject: unknown): Error =>
  subject instanceof Error ? subject : new Error(String(subject));

export const getMessageFromError = (error: Error): string => {
  if (error instanceof z.ZodError) {
    return error.issues
      .map(({ path, message }) =>
        (path.length ? [path.join("/")] : []).concat(message).join(": "),
      )
      .join("; ");
  }
  if (error instanceof OutputValidationError) {
    const hasFirstField = error.cause.issues[0]?.path.length > 0;
    return `output${hasFirstField ? "/" : ": "}${error.message}`;
  }
  return error.message;
};

/** Takes the original unvalidated examples from the properties of ZodObject schema shape */
export const pullExampleProps = <T extends z.SomeZodObject>(subject: T) =>
  Object.entries(subject.shape).reduce<Partial<z.input<T>>[]>(
    (acc, [key, schema]) => {
      const { _def } = schema as z.ZodType;
      return combinations(
        acc,
        (_def[metaSymbol]?.examples || []).map(R.objOf(key)),
        ([left, right]) => ({ ...left, ...right }),
      );
    },
    [],
  );

export const getExamples = <
  T extends z.ZodType,
  V extends "original" | "parsed" | undefined,
>({
  schema,
  variant = "original",
  validate = variant === "parsed",
  pullProps = false,
}: {
  schema: T;
  /**
   * @desc examples variant: original or parsed
   * @example "parsed" — for the case when possible schema transformations should be applied
   * @default "original"
   * @override validate: variant "parsed" activates validation as well
   * */
  variant?: V;
  /**
   * @desc filters out the examples that do not match the schema
   * @default variant === "parsed"
   * */
  validate?: boolean;
  /**
   * @desc should pull examples from properties — applicable to ZodObject only
   * @default false
   * */
  pullProps?: boolean;
}): ReadonlyArray<V extends "parsed" ? z.output<T> : z.input<T>> => {
  let examples = schema._def[metaSymbol]?.examples || [];
  if (!examples.length && pullProps && schema instanceof z.ZodObject)
    examples = pullExampleProps(schema);
  if (!validate && variant === "original") return examples;
  const result: Array<z.input<T> | z.output<T>> = [];
  for (const example of examples) {
    const parsedExample = schema.safeParse(example);
    if (parsedExample.success)
      result.push(variant === "parsed" ? parsedExample.data : example);
  }
  return result;
};

export const combinations = <T>(
  a: T[],
  b: T[],
  merge: (pair: [T, T]) => T,
): T[] => (a.length && b.length ? R.xprod(a, b).map(merge) : a.concat(b));

/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const makeCleanId = (...args: string[]) => {
  const byAlpha = R.chain((entry) => entry.split(/[^A-Z0-9]/gi), args);
  const byWord = R.chain(
    (entry) =>
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    byAlpha,
  );
  return byWord.map(ucFirst).join("");
};

export const getTransformedType = R.tryCatch(
  <T>(schema: z.ZodEffects<z.ZodTypeAny, unknown, T>, sample: T) =>
    typeof schema.parse(sample),
  R.always(undefined),
);

/** @desc can still be an array, use Array.isArray() or rather R.type() to exclude that case */
export const isObject = (subject: unknown) =>
  typeof subject === "object" && subject !== null;

export const isProduction = R.memoizeWith(
  () => process.env.TSUP_STATIC as string, // eslint-disable-line no-restricted-syntax -- substituted by TSUP
  () => process.env.NODE_ENV === "production", // eslint-disable-line no-restricted-syntax -- memoized
);
