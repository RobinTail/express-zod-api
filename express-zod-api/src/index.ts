import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
