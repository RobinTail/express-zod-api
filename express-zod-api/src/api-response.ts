import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
