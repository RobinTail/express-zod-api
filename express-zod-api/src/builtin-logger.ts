import ansis from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import { FlatObject, isProduction } from "./common-helpers";
import {
  AbstractLogger,
  formatDuration,
  isHidden,
  Severity,
  styles,
} from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /** @desc Enables colors on printed severity and inspected entities */
  color: boolean;
  /**
   * @desc Printing logs does not delay execution
   * @see https://expressjs.com/en/advanced/best-practice-performance.html#do-logging-correctly
   * */
  async: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @example null
   * @example Infinity
   * */
  depth: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: Severity | ((ms: number) => Severity);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected readonly config: BuiltinLoggerConfig;
  protected readonly stack: string[] = [];
  protected postponed?: NodeJS.Immediate;

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor({
    color = ansis.isSupported(),
    level = isProduction() ? "warn" : "debug",
    async = isProduction(),
    depth = 2,
    ctx = {},
  }: Partial<BuiltinLoggerConfig> = {}) {
    this.config = { color, level, async, depth, ctx };
  }

  protected format(subject: unknown) {
    const { depth, color: colors, level } = this.config;
    return inspect(subject, {
      depth,
      colors,
      breakLength: level === "debug" ? 80 : Infinity,
      compact: level === "debug" ? 3 : true,
    });
  }

  protected print(method: Severity, message: string, meta?: unknown) {
    const {
      level,
      ctx: { requestId, ...ctx },
      color: hasColor,
      async: isAsync,
    } = this.config;
    if (level === "silent" || isHidden(method, level)) return;
    const output: string[] = [new Date().toISOString()];
    if (requestId) output.push(hasColor ? styles.ctx(requestId) : requestId);
    output.push(
      hasColor ? `${styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) output.push(this.format(meta));
    if (Object.keys(ctx).length > 0) output.push(this.format(ctx));
    (isAsync ? this.postpone.bind(this) : console.log)(output.join(" "));
  }

  protected postpone(line: string) {
    this.stack.push(line);
    clearImmediate(this.postponed);
    this.postponed = setImmediate(this.purge.bind(this)).unref(); // do not block process.exit()
  }

  /** @internal */
  public purge() {
    while (this.stack.length) console.log(this.stack.shift());
  }

  /** @internal */
  public dispose(cb: () => void) {
    this.config.level = "silent"; // no more logs
    clearImmediate(this.postponed);
    this.postponed = setImmediate(() => {
      this.purge();
      cb();
    });
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  /**
   * @desc The argument used for instance created by .child() method
   * @see ChildLoggerProvider
   * */
  public get ctx() {
    return this.config.ctx;
  }

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = formatDuration,
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
