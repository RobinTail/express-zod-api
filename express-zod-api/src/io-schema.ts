import * as R from "ramda";
import { z } from "zod/v4";
import { AbstractMiddleware } from "./middleware";

type Base = object & { [Symbol.iterator]?: never };

/** @desc The type allowed on the top level of Middlewares and Endpoints */
export type IOSchema = z.ZodType<Base>;

/**
 * @description intersects input schemas of middlewares and the endpoint
 * @since 07.03.2022 former combineEndpointAndMiddlewareInputSchemas()
 * @since 05.03.2023 is immutable to metadata
 * @since 26.05.2024 uses the regular ZodIntersection
 * @since 22.05.2025 does not mix examples in after switching to Zod 4
 */
export const getFinalEndpointInputSchema = <
  MIN extends IOSchema,
  IN extends IOSchema,
>(
  middlewares: AbstractMiddleware[],
  input: IN,
) =>
  R.pluck("schema", middlewares)
    .concat(input)
    .reduce((acc, schema) => acc.and(schema)) as z.ZodIntersection<MIN, IN>;
