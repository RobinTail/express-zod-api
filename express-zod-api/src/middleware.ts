import { NextFunction, Request, Response } from "express";
import { z } from "zod";
import { EmptySchema, FlatObject } from "./common-helpers";
import { InputValidationError } from "./errors";
import { IOSchema } from "./io-schema";
import { LogicalContainer } from "./logical-container";
import { Security } from "./security";
import { ActualLogger } from "./logger-helpers";

type Handler<IN, OPT, OUT> = (params: {
  input: IN;
  options: OPT;
  request: Request;
  response: Response;
  logger: ActualLogger;
}) => Promise<OUT>;

export abstract class AbstractMiddleware {
  public abstract getSecurity(): LogicalContainer<Security> | undefined;
  public abstract getSchema(): IOSchema<"strip">;
  public abstract execute(params: {
    input: unknown;
    options: FlatObject;
    request: Request;
    response: Response;
    logger: ActualLogger;
  }): Promise<FlatObject>;
}

export class Middleware<
  OPT extends FlatObject,
  OUT extends FlatObject,
  SCO extends string,
  IN extends IOSchema<"strip"> = EmptySchema,
> extends AbstractMiddleware {
  readonly #schema: IN;
  readonly #security?: LogicalContainer<
    Security<Extract<keyof z.input<IN>, string>, SCO>
  >;
  readonly #handler: Handler<z.output<IN>, OPT, OUT>;

  constructor({
    input = z.object({}) as IN,
    security,
    handler,
  }: {
    input?: IN;
    security?: LogicalContainer<
      Security<Extract<keyof z.input<IN>, string>, SCO>
    >;
    handler: Handler<z.output<IN>, OPT, OUT>;
  }) {
    super();
    this.#schema = input;
    this.#security = security;
    this.#handler = handler;
  }

  public override getSecurity() {
    return this.#security;
  }

  public override getSchema() {
    return this.#schema;
  }

  /** @throws InputValidationError */
  public override async execute({
    input,
    ...rest
  }: {
    input: unknown;
    options: OPT;
    request: Request;
    response: Response;
    logger: ActualLogger;
  }) {
    try {
      const validInput = (await this.#schema.parseAsync(input)) as z.output<IN>;
      return this.#handler({ ...rest, input: validInput });
    } catch (e) {
      throw e instanceof z.ZodError ? new InputValidationError(e) : e;
    }
  }
}

export class ExpressMiddleware<
  R extends Request,
  S extends Response,
  OUT extends FlatObject,
> extends Middleware<FlatObject, OUT, string> {
  constructor(
    nativeMw: (
      request: R,
      response: S,
      next: NextFunction,
    ) => void | Promise<void>,
    {
      provider = () => ({}) as OUT,
      transformer = (err: Error) => err,
    }: {
      provider?: (request: R, response: S) => OUT | Promise<OUT>;
      transformer?: (err: Error) => Error;
    } = {},
  ) {
    super({
      handler: async ({ request, response }) =>
        new Promise<OUT>((resolve, reject) => {
          const next = (err?: unknown) => {
            if (err && err instanceof Error) return reject(transformer(err));
            resolve(provider(request as R, response as S));
          };
          nativeMw(request as R, response as S, next)?.catch(next);
        }),
    });
  }
}
