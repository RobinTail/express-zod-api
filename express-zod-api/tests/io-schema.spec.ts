import { expectTypeOf } from "vitest";
import { z } from "zod/v4";
import { IOSchema, Middleware, ez } from "../src";
import { getFinalEndpointInputSchema } from "../src/io-schema";
import { AbstractMiddleware } from "../src/middleware";

describe("I/O Schema and related helpers", () => {
  describe("IOSchema", () => {
    test("accepts object", () => {
      expectTypeOf(z.object({})).toExtend<IOSchema>();
      expectTypeOf(z.object({}).strip()).toExtend<IOSchema>();
      expectTypeOf(z.object({}).strict()).toExtend<IOSchema>();
      expectTypeOf(z.object({}).loose()).toExtend<IOSchema>();
    });
    test("accepts ez.raw()", () => {
      expectTypeOf(ez.raw()).toExtend<IOSchema>();
      expectTypeOf(ez.raw({ something: z.any() })).toExtend<IOSchema>();
    });
    test("accepts ez.form()", () => {
      expectTypeOf(ez.form({})).toExtend<IOSchema>();
    });
    test("accepts union of objects", () => {
      expectTypeOf(z.union([z.object({}), z.object({})])).toExtend<IOSchema>();
      expectTypeOf(z.object({}).or(z.object({}))).toExtend<IOSchema>();
      expectTypeOf(
        z.object({}).or(z.object({}).or(z.object({}))),
      ).toExtend<IOSchema>();
    });
    test("does not accept union of object and array of objects", () => {
      expectTypeOf(
        z.object({}).or(z.array(z.object({}))),
      ).not.toExtend<IOSchema>();
    });
    test("accepts intersection of objects", () => {
      expectTypeOf(z.object({}).and(z.object({}))).toExtend<IOSchema>();
      expectTypeOf(z.object({}).and(z.object({}))).toExtend<IOSchema>();
      expectTypeOf(
        z.object({}).and(z.object({})).and(z.object({})),
      ).toExtend<IOSchema>();
    });
    test("does not accepts intersection of object with array of objects", () => {
      expectTypeOf(
        z.object({}).and(z.array(z.object({}))),
      ).not.toExtend<IOSchema>();
    });
    test("accepts discriminated union of objects", () => {
      expectTypeOf(
        z.discriminatedUnion("type", [
          z.object({ type: z.literal("one") }),
          z.object({ type: z.literal("two") }),
        ]),
      ).toExtend<IOSchema>();
    });
    test("accepts a mix of types based on object", () => {
      expectTypeOf(
        z.object({}).or(z.object({}).and(z.object({}))),
      ).toExtend<IOSchema>();
      expectTypeOf(
        z.object({}).and(z.object({}).or(z.object({}))),
      ).toExtend<IOSchema>();
    });
    describe("Feature #600: Top level refinements", () => {
      test("accepts a refinement of object", () => {
        expectTypeOf(z.object({}).refine(() => true)).toExtend<IOSchema>();
        expectTypeOf(z.object({}).check(() => void 0)).toExtend<IOSchema>();
        expectTypeOf(
          z.object({}).refine(() => true, { error: "test" }),
        ).toExtend<IOSchema>();
      });
      test("Issue 662: accepts nested refinements", () => {
        expectTypeOf(
          z
            .object({})
            .refine(() => true)
            .refine(() => true)
            .refine(() => true),
        ).toExtend<IOSchema>();
        expectTypeOf(
          z
            .object({})
            .check(() => void 0)
            .check(() => void 0)
            .check(() => void 0),
        ).toExtend<IOSchema>();
      });
    });
    describe("Feature #1869: Top level transformations", () => {
      test("accepts transformations to another object", () => {
        expectTypeOf(
          z.object({ s: z.string() }).transform(() => ({ n: 123 })),
        ).toExtend<IOSchema>();
      });
      test("accepts nested transformations", () => {
        expectTypeOf(
          z
            .object({ s: z.string() })
            .transform(() => ({ a: 123 }))
            .transform(() => ({ b: 456 }))
            .transform(() => ({ c: 789 })),
        ).toExtend<IOSchema>();
      });
      test("accepts piping into another object schema", () => {
        expectTypeOf(
          z
            .object({ s: z.string() })
            .transform(() => ({ n: 123 }))
            .pipe(z.object({ n: z.number() })),
        ).toExtend<IOSchema>();
        expectTypeOf(
          z.object({ user_id: z.string() }).remap({ user_id: "userId" }),
        ).toExtend<IOSchema>();
      });
      test("does not accept transformation to another type", () => {
        expectTypeOf(
          z.object({}).transform(() => true),
        ).not.toExtend<IOSchema>();
        expectTypeOf(z.object({}).transform(String)).not.toExtend<IOSchema>();
        expectTypeOf(z.object({}).transform(() => [])).not.toExtend<IOSchema>();
      });
      test("does not accept piping into another kind of schema", () => {
        expectTypeOf(z.unknown().pipe(z.string())).not.toExtend<IOSchema>();
        expectTypeOf(
          z
            .object({ s: z.string() })
            .transform(Object.values)
            .pipe(z.array(z.string())),
        ).not.toExtend<IOSchema>();
      });
    });
  });

  describe("getFinalEndpointInputSchema()", () => {
    test("Should handle no middlewares", () => {
      const middlewares: AbstractMiddleware[] = [];
      const endpointInput = z.object({
        four: z.boolean(),
      });
      const result = getFinalEndpointInputSchema(middlewares, endpointInput);
      expect(result).toBeInstanceOf(z.ZodObject);
      expect(result).toMatchSnapshot();
    });

    test("Should merge input object schemas", () => {
      const middlewares: AbstractMiddleware[] = [
        new Middleware({
          input: z.object({ one: z.string() }),
          handler: vi.fn(),
        }),
        new Middleware({
          input: z.object({ two: z.number() }),
          handler: vi.fn(),
        }),
        new Middleware({
          input: z.object({ three: z.null() }),
          handler: vi.fn(),
        }),
      ];
      const endpointInput = z.object({
        four: z.boolean(),
      });
      const result = getFinalEndpointInputSchema(middlewares, endpointInput);
      expect(result).toBeInstanceOf(z.ZodIntersection);
      expect(result).toMatchSnapshot();
    });

    test("Should merge union object schemas", () => {
      const middlewares: AbstractMiddleware[] = [
        new Middleware({
          input: z
            .object({ one: z.string() })
            .or(z.object({ two: z.number() })),
          handler: vi.fn(),
        }),
        new Middleware({
          input: z
            .object({ three: z.null() })
            .or(z.object({ four: z.boolean() })),
          handler: vi.fn(),
        }),
      ];
      const endpointInput = z
        .object({ five: z.string() })
        .or(z.object({ six: z.number() }));
      const result = getFinalEndpointInputSchema(middlewares, endpointInput);
      expect(result).toBeInstanceOf(z.ZodIntersection);
      expect(result).toMatchSnapshot();
    });

    test("Should merge intersection object schemas", () => {
      const middlewares: AbstractMiddleware[] = [
        new Middleware({
          input: z
            .object({ one: z.string() })
            .and(z.object({ two: z.number() })),
          handler: vi.fn(),
        }),
        new Middleware({
          input: z
            .object({ three: z.null() })
            .and(z.object({ four: z.boolean() })),
          handler: vi.fn(),
        }),
      ];
      const endpointInput = z
        .object({ five: z.string() })
        .and(z.object({ six: z.number() }));
      const result = getFinalEndpointInputSchema(middlewares, endpointInput);
      expect(result).toBeInstanceOf(z.ZodIntersection);
      expect(result).toMatchSnapshot();
    });

    test("Zod Issue #600: can not intersect object schema with passthrough and transformation", () => {
      // @see https://github.com/colinhacks/zod/issues/600
      // Limitation: IOSchema in middlewares must be of "strip" kind
      const left = z.looseObject({});
      const right = z.object({
        id: z.string().transform((str) => parseInt(str)),
      });
      const schema = left.and(right);
      expect(() => schema.parse({ id: "123" })).toThrowErrorMatchingSnapshot();
    });

    test("Should merge mixed object schemas", () => {
      const middlewares: AbstractMiddleware[] = [
        new Middleware({
          input: z
            .object({ one: z.string() })
            .and(z.object({ two: z.number() })),
          handler: vi.fn(),
        }),
        new Middleware({
          input: z
            .object({ three: z.null() })
            .or(z.object({ four: z.boolean() })),
          handler: vi.fn(),
        }),
      ];
      const endpointInput = z.object({
        five: z.string(),
      });
      const result = getFinalEndpointInputSchema(middlewares, endpointInput);
      expect(result).toBeInstanceOf(z.ZodIntersection);
      expect(result).toMatchSnapshot();
    });

    test("Should merge examples", () => {
      const middlewares: AbstractMiddleware[] = [
        new Middleware({
          input: z
            .object({ one: z.string() })
            .and(z.object({ two: z.number() }))
            .example({ one: "test", two: 123 }),
          handler: vi.fn(),
        }),
        new Middleware({
          input: z
            .object({ three: z.null() })
            .or(z.object({ four: z.boolean() }))
            .example({ three: null, four: true }),
          handler: vi.fn(),
        }),
      ];
      const endpointInput = z
        .object({ five: z.string() })
        .example({ five: "some" });
      const result = getFinalEndpointInputSchema(middlewares, endpointInput);
      expect(result.meta()?.examples).toEqual([
        {
          one: "test",
          two: 123,
          three: null,
          four: true,
          five: "some",
        },
      ]);
    });
  });
});
