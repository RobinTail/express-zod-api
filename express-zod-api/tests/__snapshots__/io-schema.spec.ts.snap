// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`I/O Schema and related helpers > getFinalEndpointInputSchema() > Should handle no middlewares 1`] = `
{
  "_type": "ZodObject",
  "shape": {
    "four": {
      "_type": "ZodBoolean",
    },
  },
}
`;

exports[`I/O Schema and related helpers > getFinalEndpointInputSchema() > Should merge input object schemas 1`] = `
{
  "_type": "ZodIntersection",
  "left": {
    "_type": "ZodIntersection",
    "left": {
      "_type": "ZodIntersection",
      "left": {
        "_type": "ZodObject",
        "shape": {
          "one": {
            "_type": "ZodString",
          },
        },
      },
      "right": {
        "_type": "ZodObject",
        "shape": {
          "two": {
            "_type": "ZodNumber",
          },
        },
      },
    },
    "right": {
      "_type": "ZodObject",
      "shape": {
        "three": {
          "_type": "ZodNull",
        },
      },
    },
  },
  "right": {
    "_type": "ZodObject",
    "shape": {
      "four": {
        "_type": "ZodBoolean",
      },
    },
  },
}
`;

exports[`I/O Schema and related helpers > getFinalEndpointInputSchema() > Should merge intersection object schemas 1`] = `
{
  "_type": "ZodIntersection",
  "left": {
    "_type": "ZodIntersection",
    "left": {
      "_type": "ZodIntersection",
      "left": {
        "_type": "ZodObject",
        "shape": {
          "one": {
            "_type": "ZodString",
          },
        },
      },
      "right": {
        "_type": "ZodObject",
        "shape": {
          "two": {
            "_type": "ZodNumber",
          },
        },
      },
    },
    "right": {
      "_type": "ZodIntersection",
      "left": {
        "_type": "ZodObject",
        "shape": {
          "three": {
            "_type": "ZodNull",
          },
        },
      },
      "right": {
        "_type": "ZodObject",
        "shape": {
          "four": {
            "_type": "ZodBoolean",
          },
        },
      },
    },
  },
  "right": {
    "_type": "ZodIntersection",
    "left": {
      "_type": "ZodObject",
      "shape": {
        "five": {
          "_type": "ZodString",
        },
      },
    },
    "right": {
      "_type": "ZodObject",
      "shape": {
        "six": {
          "_type": "ZodNumber",
        },
      },
    },
  },
}
`;

exports[`I/O Schema and related helpers > getFinalEndpointInputSchema() > Should merge mixed object schemas 1`] = `
{
  "_type": "ZodIntersection",
  "left": {
    "_type": "ZodIntersection",
    "left": {
      "_type": "ZodIntersection",
      "left": {
        "_type": "ZodObject",
        "shape": {
          "one": {
            "_type": "ZodString",
          },
        },
      },
      "right": {
        "_type": "ZodObject",
        "shape": {
          "two": {
            "_type": "ZodNumber",
          },
        },
      },
    },
    "right": {
      "_type": "ZodUnion",
      "options": [
        {
          "_type": "ZodObject",
          "shape": {
            "three": {
              "_type": "ZodNull",
            },
          },
        },
        {
          "_type": "ZodObject",
          "shape": {
            "four": {
              "_type": "ZodBoolean",
            },
          },
        },
      ],
    },
  },
  "right": {
    "_type": "ZodObject",
    "shape": {
      "five": {
        "_type": "ZodString",
      },
    },
  },
}
`;

exports[`I/O Schema and related helpers > getFinalEndpointInputSchema() > Should merge union object schemas 1`] = `
{
  "_type": "ZodIntersection",
  "left": {
    "_type": "ZodIntersection",
    "left": {
      "_type": "ZodUnion",
      "options": [
        {
          "_type": "ZodObject",
          "shape": {
            "one": {
              "_type": "ZodString",
            },
          },
        },
        {
          "_type": "ZodObject",
          "shape": {
            "two": {
              "_type": "ZodNumber",
            },
          },
        },
      ],
    },
    "right": {
      "_type": "ZodUnion",
      "options": [
        {
          "_type": "ZodObject",
          "shape": {
            "three": {
              "_type": "ZodNull",
            },
          },
        },
        {
          "_type": "ZodObject",
          "shape": {
            "four": {
              "_type": "ZodBoolean",
            },
          },
        },
      ],
    },
  },
  "right": {
    "_type": "ZodUnion",
    "options": [
      {
        "_type": "ZodObject",
        "shape": {
          "five": {
            "_type": "ZodString",
          },
        },
      },
      {
        "_type": "ZodObject",
        "shape": {
          "six": {
            "_type": "ZodNumber",
          },
        },
      },
    ],
  },
}
`;

exports[`I/O Schema and related helpers > getFinalEndpointInputSchema() > Zod Issue #600: can not intersect object schema with passthrough and transformation 1`] = `
ZodError({
  "message": "[
  {
    "code": "invalid_intersection_types",
    "path": [],
    "message": "Intersection results could not be merged"
  }
]",
})
`;
